### ISSUE TITLE
InferenceRequest::serialize does not handle logits post processor, log an error

### ISSUE THREAD

**Issue by DreamGenX** [NVIDIA/TensorRT-LLM#1771]
### System Info

When using C++ GptManager, I was using InferenceRequest::serialize to synchronize requests between ranks, not realizing that it does not handle logits post processors.

It would be great to log an error in InferenceRequest::serialize when logits post processor is not nullopt. Or, at the very least, document it in [code](https://github.com/NVIDIA/TensorRT-LLM/blob/main/cpp/include/tensorrt_llm/batch_manager/inferenceRequest.h) and the [readme](https://github.com/NVIDIA/TensorRT-LLM/blob/db4edea1e1359bcfcac7bbb87c1b639b5611c721/docs/source/advanced/batch-manager.md?plain=1#L125)

Here are some issues this resulted in:
- https://github.com/NVIDIA/TensorRT-LLM/issues/1767 (my own :))
- https://github.com/NVIDIA/TensorRT-LLM/issues/1415

### Who can help?

@ju

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

N/A

### Expected behavior

N/A

### actual behavior

N/A

### additional notes

N/A

**Comment by nv-guomingz** [comment]
Hi @MartinMarciniszyn  would u plz take a look this ticket?

**Comment by MartinMarciniszyn** [comment]
Hi @DreamGenX, have you tried using the new Executor API instead? Requests of this API will serialize the name of the associated logits post-processor correctly.

**Comment by DreamGenX** [comment]
Hi @MartinMarciniszyn, thanks for your time. Comment on that below, but I would say that adding documentation and runtime warning for batch manager's `InferenceRequest::serialize` still makes sense, regardless of the viability of Executor API.

**Regarding Executor API:**

Executor API is unfortunately limited as far as I can tell -- I don't see a way to have "dynamic" post-processors where the post-processors differ per request and all possible valuas aren't know apriori (so can't be registred in the map).

With the executor API, since all post-processors need to be pre-registered, there's no away to achieve this as far as I can tell. We can't even store the state in some external storage keyed by `requestId`, since the id isn't know before calling `enqueueRequests`.

_Possible solution:_ Let `LogitPostProcessor` access arbitrary `std::vector<uint8_t> state` that we provide when building the `Request` / `SamplingParams`, there we can encode additional per-request state for the logit post processor. A more flexible version would be to actually have `LogitPostProcessorFactory` that returns `LogitPostProcessor` based on the `state`, as it could avoid expensive pre-processing on each token in some cases.

This is the main reason I have essentially re-implemented the Executor C++ API myself (the code is not open-source, so I could not just modify it either). So if the Executor API can be made more flexible, I would love it.

See also here: https://github.com/NVIDIA/TensorRT-LLM/issues/1680#issuecomment-2137826264

**Comment by akhoroshev** [comment]
> This is the main reason I have essentially re-implemented the Executor C++ API myself (the code is not open-source, so I could not just modify it either). So if the Executor API can be made more flexible, I would love it.

@DreamGenX 

Hi, do you have any plans to open source the your own "Executor"?

**Comment by DreamGenX** [comment]
Hi @akhoroshev -- I would hope it soon becomes obsolete once the Executor API becomes more mature, but if not I don't mind releasing it.

**Comment by akhoroshev** [comment]
> Hi @akhoroshev -- I would hope it soon becomes obsolete once the Executor API becomes more mature, but if not I don't mind releasing it.

@DreamGenX 

The trtllm team's core idea of ​​not opening the source code (Executor/BatchManager) makes it very difficult to customize functionality.

There is a list of features that cannot be implemented due to closed sources.
* https://github.com/NVIDIA/TensorRT-LLM/issues/1681
* https://github.com/NVIDIA/TensorRT-LLM/issues/1654
* https://github.com/NVIDIA/TensorRT-LLM/issues/595

Therefore, I would be very grateful if you could share the sources of your executor.

**Comment by DreamGenX** [comment]
@MartinMarciniszyn can you let us know if you plan to expose more flexible API for logit post processors in the Executor?

**Comment by MartinMarciniszyn** [comment]
Thank you for your feedback, @DreamGenX.

I agree that there should an error message for `InferenceRequest::serialize`. We will look into it.

Serializing arbitrary function callbacks between the different ranks poses technical difficulties. The team will investigate how we can add more flexibility to the API given these restrictions.

**Comment by DreamGenX** [comment]
@MartinMarciniszyn Yeah, of course, serialization of arbitrary callbacks is not feasible, that's why I propose that we can attach arbitrary binary data as extension to the Request / SamplingParams, and then get access to this data in the LogitPostProcessor. It would be even better if we can have LogitPostProcessorFactory, which gets called at the start of the request and can do all the pre-processing work (e.g. deserialization of the data, etc.) only once, instead of on every token.

**Comment by nv-guomingz** [comment]
Hi @DreamGenX do u still have further issue or question now? If not, we'll close it soon.

**Comment by MartinMarciniszyn** [comment]
We have made some architectural changes so that the logits post-processor will always execute on rank 0. This should also resolve the serialization issues.
